tsfc/coffee_mode.py

```diff
@@ -0,0 +1,338 @@
+                if atomic not in atomic_index:
+                    atomic_index[atomic] = next(index)
+                connection.append(atomic_index[atomic])
+            connections.append(tuple(connection))
+
+    if len(atomic_index) == 0:
+        return ((), ())
+    if len(atomic_index) == 1:
+        return ((next(iterkeys(atomic_index)), ), ())
+
+    # set up the ILP
+    import pulp as ilp
+    ilp_prob = ilp.LpProblem('gem factorise', ilp.LpMinimize)
+    ilp_var = ilp.LpVariable.dicts('node', range(len(atomic_index)), 0, 1, ilp.LpBinary)
+
+    # Objective function
+    # Minimise number of factors to pull. If same number, favour factor with larger extent
+    big = 1e20  # some arbitrary big number
+    ilp_prob += ilp.lpSum(ilp_var[index] * (big - index_extent(atomic, argument_indices)) for atomic, index in iteritems(atomic_index))
+
+    # constraints
+    for connection in connections:
+        ilp_prob += ilp.lpSum(ilp_var[index] for index in connection) >= 1
+
+    ilp_prob.solve()
+    if ilp_prob.status != 1:
+        raise RuntimeError("Something bad happened during ILP")
+
+    def optimal(atomic):
+        return ilp_var[atomic_index[atomic]].value() == 1
+
+    optimal_atomics = filter(optimal, iterkeys(atomic_index))
+    other_atomics = filterfalse(optimal, iterkeys(atomic_index))
+
+    return (tuple(optimal_atomics), tuple(other_atomics))
+
+
+def factorise_atomics(monomial_sum, optimal_atomics, argument_indices):
+    """Group and factorise monomials using a list of atomics as common
+    subexpressions. Create new monomials for each group and optimise them recursively.
+
+    :arg monomial_sum: a :class:`MonomialSum` object
+    :arg optimal_atomics: list of tuples of atomics to be used as common subexpression
+                          and the frozenset of their sum indices
+    :arg argument_indices: tuple of argument indices
+
+    :returns: a factorised :class:`MonomialSum` object, or the original object
+    if no changes are made
+    """
+    if not optimal_atomics:
+        return monomial_sum
+    if len(monomial_sum.ordering) < 2:
+        return monomial_sum
+    new_monomial_sum = MonomialSum()
+    # Group monomials with respect to each optimal atomic
+    factor_group = OrderedDict()
+    for monomial in monomial_sum:
+        for sum_indices, oa in optimal_atomics:
+            if frozenset(monomial.sum_indices) == frozenset(sum_indices) and oa in monomial.atomics:
+                # Add monomial to the list of corresponding optimal atomic
+                factor_group.setdefault((sum_indices, oa), []).append(monomial)
+                break
+        else:
+            # Add monomials that do no have argument factors to new MonomialSum
+            new_monomial_sum.add(monomial.sum_indices, monomial.atomics, monomial.rest)
+    # We should not drop monomials
+    assert sum(map(len, itervalues(factor_group))) + len(list(new_monomial_sum)) == len(list(monomial_sum))
```

Comment:
You shall replace `len(list(...))` with `len(...)` once `__len__` is added to `MonomialSum`.